plugins/filter/find_challenges.py

```diff
@@ -0,0 +1,84 @@
+from ansible.errors import AnsibleError
+
+
+def find_challenges(challenge_response: dict, challenge_type: str, expected_domains: list) -> dict:
+    challenges = build_challenges(challenge_response, challenge_type)
+
+    if set(challenges.keys()) == set(expected_domains):
+        return challenges
+    else:
+        errors = []
+        expected_not_found = set(expected_domains).difference(set(challenges.keys()))
+        if expected_not_found:
+            errors.append(f"Expected {challenge_type} challenges for the following domains not found: "
+                          f"{','.join(expected_not_found)}")
+
+        unexpected_from_api = set(challenges.keys()).difference(set(expected_domains))
+        if unexpected_from_api:
+            errors.append(f"The API responded with {challenge_type} challenges for the following domains "
+                          f"we did not expect: {','.join(unexpected_from_api)}")
+
+        if errors:
+            raise AnsibleError(" // ".join(errors))
+
+
+def build_challenges(challenge_response: dict, challenge_type: str) -> dict:
+    if challenge_type not in ['http-01', 'dns-01']:
+        raise AssertionError(f"Unknown challenge_type ({challenge_type})")
+
+    if len(challenge_response['challenge_data']) > 0:
```

Comment:
I also got rid of the Asserts since a valid input should always contain the needed fields, It also seems that the later assertions should catch any formal errors with the returned data.